src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -0,0 +1,686 @@
+                              TraceLoggingInt64(index, "index", "the index of the hotkey to add"),
+                              TraceLoggingWideString(cmd.Name().c_str(), "name", "the name of the command"),
+                              TraceLoggingBoolean(succeeded, "succeeded", "true if we succeeded"),
+                              TraceLoggingLevel(WINEVENT_LEVEL_VERBOSE),
+                              TraceLoggingKeyword(TIL_KEYWORD_TRACE));
+            _hotkeys.emplace_back(summonArgs);
+        }
+    }
+}
+
+#pragma endregion
+////////////////////////////////////////////////////////////////////////////////
+#pragma region NotificationIcon
+// Method Description:
+// - Creates a Notification Icon and hooks up its handlers
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void WindowEmperor::_createNotificationIcon()
+{
+    _notificationIcon = std::make_unique<NotificationIcon>(_window.get());
+    _notificationIcon->SummonWindowRequested([this](auto& args) { _manager.SummonWindow(args); });
+}
+
+// Method Description:
+// - Deletes our notification icon if we have one.
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void WindowEmperor::_destroyNotificationIcon()
+{
+    _notificationIcon->RemoveIconFromNotificationArea();
+    _notificationIcon = nullptr;
+}
+
+void WindowEmperor::_checkWindowsForNotificationIcon()
+{
+    // We need to check some conditions to show the notification icon.
+    //
+    // * If there's a Quake window somewhere, we'll want to keep the
+    //   notification icon.
+    // * There's two settings - MinimizeToNotificationArea and
+    //   AlwaysShowNotificationIcon. If either one of them are true, we want to
+    //   make sure there's a notification icon.
+    //
+    // If both are false, we want to remove our icon from the notification area.
+    // When we remove our icon from the notification area, we'll also want to
+    // re-summon any hidden windows, but right now we're not keeping track of
+    // who's hidden, so just summon them all. Tracking the work to do a "summon
+    // all minimized" in GH#10448
+
+    bool needsIcon = false;
+    for (const auto& _windowThread : _windows)
+    {
+        needsIcon |= _windowThread->Logic().RequestsTrayIcon();
+    }
+
+    if (needsIcon)
+    {
+        _showNotificationIconRequested();
+    }
+    else
+    {
+        _hideNotificationIconRequested();
+    }
+}
+
+void WindowEmperor::_showNotificationIconRequested()
+{
+    if (!_notificationIcon)
+    {
+        _createNotificationIcon();
+    }
+}
+
+void WindowEmperor::_hideNotificationIconRequested()
+{
+    // Destroy it only if our settings allow it
+    if (_notificationIcon)
+    {
+        // If we no longer want the tray icon, but we did have one, then quick
+        // re-summon all our windows, so they don't get lost when the icon
+        // disappears forever.
+        _manager.SummonAllWindows();

```

Comment:
Yea, I think they all pop back up. This was in `AppHost::_HandleSettingsChanged` before:

```c++
            else if (_notificationIcon && !_logic.GetMinimizeToNotificationArea() && !_logic.GetAlwaysShowNotificationIcon())
            {
                _windowManager.SummonAllWindows();
                _DestroyNotificationIcon();
            }
```

I think the point was that if there were hidden windows and you turn off the tray icon, then they all come back instead of being hidden forever

Reply:
I hate that, but it doesn't really happen and we did it before, so meh haha.